src/main/java/io/reactivex/internal/operators/maybe/MaybeFlatMapSingle.java

```diff
@@ -0,0 +1,133 @@
+import io.reactivex.internal.disposables.DisposableHelper;
+import io.reactivex.internal.functions.ObjectHelper;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * Maps the success value of the source MaybeSource into a Single.
+ * @param <T>
+ */
+public final class MaybeFlatMapSingle<T> extends Single<T> {
+
+    final MaybeSource<T> source;
+
+    final Function<? super T, ? extends SingleSource<T>> mapper;
+
+    public MaybeFlatMapSingle(MaybeSource<T> source, Function<? super T, ? extends SingleSource<T>> mapper) {
+        this.source = source;
+        this.mapper = mapper;
+    }
+
+    @Override
+    protected void subscribeActual(SingleObserver<? super T> s) {
+        FlatMapSingleObserver<T> parent = new FlatMapSingleObserver<T>(s, mapper);
+        s.onSubscribe(parent);
+        source.subscribe(parent);
+    }
+
+    static final class FlatMapSingleObserver<T>
+    extends AtomicReference<Disposable>
+    implements MaybeObserver<T>, Disposable {
+
+        private static final long serialVersionUID = 4827726964688405508L;
+
+        final SingleObserver actual;
+
+        final Function<? super T, ? extends SingleSource<T>> mapper;
+
+        final SingleObserver<T> singleObserver = new SingleObserver<T>() {
+            @Override
+            public void onSubscribe(final Disposable d) {
+
+            }
+
+            @Override
+            public void onSuccess(final T value) {
+                actual.onSuccess(value);
+            }
+
+            @Override
+            public void onError(final Throwable e) {
+                actual.onError(e);
+            }
+        };
+
+        FlatMapSingleObserver(SingleObserver actual,
+                Function<? super T, ? extends SingleSource<T>> mapper) {
+            this.actual = actual;
+            this.mapper = mapper;
+        }
+
+        @Override
+        public void dispose() {
+            DisposableHelper.dispose(this);
+        }
+
+        @Override
+        public boolean isDisposed() {
+            return DisposableHelper.isDisposed(get());
+        }
+
+        @Override
+        public void onSubscribe(Disposable d) {
+            DisposableHelper.replace(this, d);
+        }
+
+        @Override
+        public void onSuccess(T value) {
+            SingleSource<T> ss;
+
+            try {
+                ss = ObjectHelper.requireNonNull(mapper.apply(value), "The mapper returned a null SingleSource");
+            } catch (Throwable ex) {
+                Exceptions.throwIfFatal(ex);
+                onError(ex);
+                return;
+            }
+
+            ss.subscribe(singleObserver);
+        }
+
+        @Override
+        public void onError(Throwable e) {
+            actual.onError(e);
+        }
+
+        @Override
+        public void onComplete() {
+            SingleSource<T> ss;
+
+            try {
+                ss = ObjectHelper.requireNonNull(mapper.apply(null), "The mapper returned a null SingleSource");
```

Comment:
Here is the question whether the mapper should be used or not. I thought it might make sense that `Maybe.empty()` should also go through `flatMapSingle`